File: Mobile/Android/app/src/main/kotlin/com/sendit/api/SendItAPI.kt
package com.sendit.api

import android.content.Context
import android.provider.Settings
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

// MARK: - Data Models

@Serializable
data class CareerData(
    val level: Int,
    val totalExperience: Int,
    val balance: Double,
    val racesCompleted: Int,
    val wins: Int,
    val bestLapTime: Double
)

@Serializable
data class RaceResult(
    val eventName: String,
    val trackName: String,
    val position: Int,
    val bestLapTime: Double,
    val prizeMoney: Double,
    val experiencePoints: Int,
    val completeDate: String
)

@Serializable
data class VehicleUpgrade(
    val name: String,
    val type: String,
    val cost: Double,
    val performanceGain: Double,
    val requiredLevel: Int
)

@Serializable
data class VehicleSetup(
    val name: String,
    val track: String,
    val bestLapTime: Double,
    val useCount: Int,
    val created: String
)

@Serializable
data class LoginResponse(
    val success: Boolean,
    val sessionToken: String,
    val message: String
)

@Serializable
data class RaceHistoryResponse(
    val success: Boolean,
    val races: List<RaceResult>
)

@Serializable
data class UpgradesResponse(
    val success: Boolean,
    val upgrades: List<VehicleUpgrade>
)

@Serializable
data class SetupsResponse(
    val success: Boolean,
    val setups: List<VehicleSetup>
)

@Serializable
data class PurchaseResponse(
    val success: Boolean,
    val message: String
)

// MARK: - API Client

class SendItAPIClient(
    private val context: Context,
    private val baseURL: String = "http://localhost:8080"
) {
    private val deviceId: String = Settings.Secure.getString(
        context.contentResolver,
        Settings.Secure.ANDROID_ID
    )

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .writeTimeout(10, TimeUnit.SECONDS)
        .build()

    private val json = Json { ignoreUnknownKeys = true }
    private val dataStore = context.dataStore

    private var sessionToken: String? = null

    init {
        // Try to restore session token
        try {
            val prefs = android.content.SharedPreferences(context)
            sessionToken = prefs.getString("sessionToken", null)
        } catch (e: Exception) {
            // Preference not available, will require login
        }
    }

    suspend fun login(): Result<String> = withContext(Dispatchers.IO) {
        try {
            val loginBody = Json.encodeToString(
                mapOf(
                    "deviceId" to deviceId,
                    "appVersion" to "1.0.0"
                )
            )

            val request = Request.Builder()
                .url("$baseURL/api/login")
                .post(loginBody.toRequestBody("application/json".toMediaType()))
                .header("Content-Type", "application/json")
                .build()

            val response = httpClient.newCall(request).execute()

            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Login failed"))
            }

            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            val loginResponse = json.decodeFromString<LoginResponse>(responseBody)

            sessionToken = loginResponse.sessionToken
            // Save session token to shared preferences
            saveSessionToken(loginResponse.sessionToken)

            Result.success(loginResponse.sessionToken)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun logout() {
        sessionToken = null
        clearSessionToken()
    }

    suspend fun fetchCareerData(): Result<CareerData> = withContext(Dispatchers.IO) {
        try {
            val response = makeAuthenticatedRequest("/api/career", "GET")
            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Failed to fetch career data"))
            }

            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            val careerData = json.decodeFromString<CareerData>(responseBody)
            Result.success(careerData)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun fetchRaceHistory(count: Int = 10): Result<List<RaceResult>> = withContext(Dispatchers.IO) {
        try {
            val response = makeAuthenticatedRequest("/api/career/races?count=$count", "GET")
            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Failed to fetch race history"))
            }

            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            val historyResponse = json.decodeFromString<RaceHistoryResponse>(responseBody)
            Result.success(historyResponse.races)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun fetchAvailableUpgrades(): Result<List<VehicleUpgrade>> = withContext(Dispatchers.IO) {
        try {
            val response = makeAuthenticatedRequest("/api/upgrades/available", "GET")
            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Failed to fetch upgrades"))
            }

            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            val upgradesResponse = json.decodeFromString<UpgradesResponse>(responseBody)
            Result.success(upgradesResponse.upgrades)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun purchaseUpgrade(upgradeName: String): Result<Boolean> = withContext(Dispatchers.IO) {
        try {
            val body = Json.encodeToString(mapOf("upgradeName" to upgradeName))
            val response = makeAuthenticatedRequest(
                "/api/upgrades/purchase",
                "POST",
                body
            )

            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Purchase failed"))
            }

            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            val purchaseResponse = json.decodeFromString<PurchaseResponse>(responseBody)
            Result.success(purchaseResponse.success)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun fetchSavedSetups(): Result<List<VehicleSetup>> = withContext(Dispatchers.IO) {
        try {
            val response = makeAuthenticatedRequest("/api/setups", "GET")
            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Failed to fetch setups"))
            }

            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            val setupsResponse = json.decodeFromString<SetupsResponse>(responseBody)
            Result.success(setupsResponse.setups)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun loadSetup(setupName: String): Result<Boolean> = withContext(Dispatchers.IO) {
        try {
            val body = Json.encodeToString(mapOf("setupName" to setupName))
            val response = makeAuthenticatedRequest(
                "/api/setups/load",
                "POST",
                body
            )

            if (!response.isSuccessful) {
                return@withContext Result.failure(Exception("Load failed"))
            }

            val responseBody = response.body?.string() ?: throw Exception("Empty response")
            val setupResponse = json.decodeFromString<PurchaseResponse>(responseBody)
            Result.success(setupResponse.success)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun isAuthenticated(): Boolean = sessionToken != null

    // MARK: - Private Helpers

    private fun makeAuthenticatedRequest(
        endpoint: String,
        method: String,
        body: String? = null
    ): okhttp3.Response {
        if (sessionToken == null) {
            throw Exception("Not authenticated")
        }

        val builder = Request.Builder()
            .url("$baseURL$endpoint")
            .header("Authorization", "Bearer $sessionToken")
            .header("Content-Type", "application/json")

        when (method) {
            "GET" -> builder.get()
            "POST" -> {
                val requestBody = body?.toRequestBody("application/json".toMediaType())
                builder.post(requestBody ?: "{}".toRequestBody("application/json".toMediaType()))
            }
        }

        val request = builder.build()
        return httpClient.newCall(request).execute()
    }

    private fun saveSessionToken(token: String) {
        try {
            val sharedPref = context.getSharedPreferences("SendIt", Context.MODE_PRIVATE)
            sharedPref.edit().putString("sessionToken", token).apply()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun clearSessionToken() {
        try {
            val sharedPref = context.getSharedPreferences("SendIt", Context.MODE_PRIVATE)
            sharedPref.edit().remove("sessionToken").apply()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}

// MARK: - Extension Functions

fun Double.formatMoney(): String {
    return "$%.0f".format(this)
}

fun Double.formatTime(): String {
    val minutes = this.toInt() / 60
    val seconds = this % 60
    return "%d:%06.3f".format(minutes, seconds)
}

fun Double.formatLapTime(): String {
    return "%.3f".format(this)
}

File: Mobile/iOS/SendItApp/SendItAPI.swift
import Foundation

// MARK: - Data Models

struct CareerData: Codable {
    let level: Int
    let totalExperience: Int
    let balance: Double
    let racesCompleted: Int
    let wins: Int
    let bestLapTime: Double
}

struct RaceResult: Codable {
    let eventName: String
    let trackName: String
    let position: Int
    let bestLapTime: Double
    let prizeMoney: Double
    let experiencePoints: Int
    let completeDate: String
}

struct VehicleUpgrade: Codable, Identifiable {
    let id = UUID()
    let name: String
    let type: String
    let cost: Double
    let performanceGain: Double
    let requiredLevel: Int

    enum CodingKeys: String, CodingKey {
        case name, type, cost, performanceGain, requiredLevel
    }
}

struct VehicleSetup: Codable, Identifiable {
    let id = UUID()
    let name: String
    let track: String
    let bestLapTime: Double
    let useCount: Int
    let created: String

    enum CodingKeys: String, CodingKey {
        case name, track, bestLapTime, useCount, created
    }
}

struct LoginResponse: Codable {
    let success: Bool
    let sessionToken: String
    let message: String
}

struct APIResponse<T: Codable>: Codable {
    let success: Bool
    let error: String?
    let data: T?
}

// MARK: - API Client

class SendItAPIClient: ObservableObject {
    @Published var careerData: CareerData?
    @Published var isAuthenticated = false
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let baseURL: URL
    private var sessionToken: String?
    private let deviceId: String
    private let session: URLSession

    init(baseURL: String = "http://localhost:8080", deviceId: String = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString) {
        self.baseURL = URL(string: baseURL)!
        self.deviceId = deviceId

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        config.timeoutIntervalForResource = 30
        self.session = URLSession(configuration: config)

        // Restore session token if available
        if let token = UserDefaults.standard.string(forKey: "sessionToken") {
            self.sessionToken = token
            self.isAuthenticated = true
        }
    }

    // MARK: - Authentication

    func login() async throws {
        isLoading = true
        defer { isLoading = false }

        let loginData = ["deviceId": deviceId, "appVersion": "1.0.0"]
        let jsonData = try JSONSerialization.data(withJSONObject: loginData)

        var request = URLRequest(url: baseURL.appendingPathComponent("/api/login"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = jsonData

        let (data, response) = try session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw APIError.invalidResponse
        }

        let loginResponse = try JSONDecoder().decode(LoginResponse.self, from: data)
        self.sessionToken = loginResponse.sessionToken

        // Save session token
        UserDefaults.standard.set(sessionToken, forKey: "sessionToken")

        DispatchQueue.main.async {
            self.isAuthenticated = true
        }
    }

    func logout() {
        sessionToken = nil
        UserDefaults.standard.removeObject(forKey: "sessionToken")

        DispatchQueue.main.async {
            self.isAuthenticated = false
        }
    }

    // MARK: - Career Data

    func fetchCareerData() async throws {
        isLoading = true
        defer { isLoading = false }

        let data = try await makeAuthenticatedRequest(
            endpoint: "/api/career",
            method: "GET",
            responseType: CareerData.self
        )

        DispatchQueue.main.async {
            self.careerData = data
        }
    }

    func fetchRaceHistory(count: Int = 10) async throws -> [RaceResult] {
        let data = try await makeAuthenticatedRequest(
            endpoint: "/api/career/races?count=\(count)",
            method: "GET",
            responseType: RaceHistoryResponse.self
        )
        return data.races
    }

    // MARK: - Vehicle Upgrades

    func fetchAvailableUpgrades() async throws -> [VehicleUpgrade] {
        let data = try await makeAuthenticatedRequest(
            endpoint: "/api/upgrades/available",
            method: "GET",
            responseType: UpgradesResponse.self
        )
        return data.upgrades
    }

    func purchaseUpgrade(name: String) async throws -> Bool {
        let request = ["upgradeName": name]
        let jsonData = try JSONSerialization.data(withJSONObject: request)

        let response = try await makeAuthenticatedJSONRequest(
            endpoint: "/api/upgrades/purchase",
            method: "POST",
            body: jsonData,
            responseType: PurchaseResponse.self
        )

        return response.success
    }

    // MARK: - Vehicle Setups

    func fetchSavedSetups() async throws -> [VehicleSetup] {
        let data = try await makeAuthenticatedRequest(
            endpoint: "/api/setups",
            method: "GET",
            responseType: SetupsResponse.self
        )
        return data.setups
    }

    func loadSetup(name: String) async throws -> Bool {
        let request = ["setupName": name]
        let jsonData = try JSONSerialization.data(withJSONObject: request)

        let response = try await makeAuthenticatedJSONRequest(
            endpoint: "/api/setups/load",
            method: "POST",
            body: jsonData,
            responseType: SetupResponse.self
        )

        return response.success
    }

    // MARK: - Helper Methods

    private func makeAuthenticatedRequest<T: Codable>(
        endpoint: String,
        method: String,
        responseType: T.Type
    ) async throws -> T {
        guard let token = sessionToken else {
            throw APIError.notAuthenticated
        }

        var request = URLRequest(url: baseURL.appendingPathComponent(endpoint))
        request.httpMethod = method
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let (data, response) = try session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        if httpResponse.statusCode == 401 {
            // Session expired
            self.logout()
            throw APIError.sessionExpired
        }

        guard httpResponse.statusCode == 200 else {
            throw APIError.serverError(httpResponse.statusCode)
        }

        return try JSONDecoder().decode(T.self, from: data)
    }

    private func makeAuthenticatedJSONRequest<T: Codable>(
        endpoint: String,
        method: String,
        body: Data,
        responseType: T.Type
    ) async throws -> T {
        guard let token = sessionToken else {
            throw APIError.notAuthenticated
        }

        var request = URLRequest(url: baseURL.appendingPathComponent(endpoint))
        request.httpMethod = method
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = body

        let (data, response) = try session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            throw APIError.serverError(httpResponse.statusCode)
        }

        return try JSONDecoder().decode(T.self, from: data)
    }
}

// MARK: - Supporting Response Types

struct RaceHistoryResponse: Codable {
    let success: Bool
    let races: [RaceResult]
}

struct UpgradesResponse: Codable {
    let success: Bool
    let upgrades: [VehicleUpgrade]
}

struct SetupsResponse: Codable {
    let success: Bool
    let setups: [VehicleSetup]
}

struct PurchaseResponse: Codable {
    let success: Bool
    let message: String
}

struct SetupResponse: Codable {
    let success: Bool
    let message: String
}

// MARK: - Error Types

enum APIError: LocalizedError {
    case invalidResponse
    case notAuthenticated
    case sessionExpired
    case serverError(Int)
    case decodingError
    case networkError(String)

    var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Invalid server response"
        case .notAuthenticated:
            return "Not authenticated. Please login first."
        case .sessionExpired:
            return "Your session has expired. Please login again."
        case .serverError(let code):
            return "Server error: \(code)"
        case .decodingError:
            return "Failed to decode response"
        case .networkError(let message):
            return "Network error: \(message)"
        }
    }
}

// MARK: - Utility Extensions

extension Double {
    func formatMoney() -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencySymbol = "$"
        return formatter.string(from: NSNumber(value: self)) ?? "$0"
    }

    func formatTime() -> String {
        let minutes = Int(self) / 60
        let seconds = self.truncatingRemainder(dividingBy: 60)
        return String(format: "%d:%06.3f", minutes, seconds)
    }

    func formatLapTime() -> String {
        return String(format: "%.3f", self)
    }
}
